eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxInfo.hs

```diff
@@ -542,89 +549,187 @@ instance Semigroup ScriptResult where
+      pdSBS :: SBS.ShortByteString,
+      pdPlutusData :: PlutusData l,
+      pdProtVer :: ProtVer
+    } ->
+    PlutusDebug l
 
 -- | There is dummy Show instance for PlutusDebug intentionally, because it is too
 -- expensive and it will be too tempting to use it incorrectly. If needed for
 -- testing use 'StandaloneDeriving', otherwise define an efficient way to display
 -- this info.
-instance Show PlutusDebug where
+instance Show (PlutusDebug l) where
   show _ = "PlutusDebug Omitted"
 
-data PlutusError = PlutusErrorV1 PV1.EvaluationError | PlutusErrorV2 PV2.EvaluationError
+deriving instance Eq (PlutusDebug l)
+
+deriving instance Generic (PlutusDebug l)
+
+mkPlutusDebug ::
+  forall (l :: Language).
+  SLanguage l ->
+  CostModel ->
+  ExUnits ->
+  SBS.ShortByteString ->
+  [PCD.Data] ->
+  ProtVer ->
+  PlutusDebug l
+mkPlutusDebug sl costModel exUnits sbs pData protVer = case sl of
+  SPlutusV1 -> PlutusDebug costModel exUnits sbs (DataV1 pData) protVer
+  SPlutusV2 -> PlutusDebug costModel exUnits sbs (DataV2 pData) protVer
+
+plutusDebugLang ::
+  forall (l :: Language).
+  (IsLanguage l) =>
+  PlutusDebug l ->
+  (SLanguage l, PlutusDebug l)
+plutusDebugLang p = (isLanguage, p)
+
+instance
+  forall (l :: Language).
+  (Typeable l, IsLanguage l) =>
+  ToCBOR (PlutusDebug l)
+  where
+  toCBOR (PlutusDebug costModel exUnits sbs pData protVer) = case pData of
+    DataV1 _ ->
+      encode $ Sum PlutusDebug 0 !> E encodeCostModel costModel !> To exUnits !> To sbs !> To pData !> To protVer
+    DataV2 _ ->
+      encode $ Sum PlutusDebug 1 !> E encodeCostModel costModel !> To exUnits !> To sbs !> To pData !> To protVer
+
+instance
+  forall (l :: Language).
+  (Typeable l, IsLanguage l) =>
+  FromCBOR (PlutusDebug l)
+  where
+  fromCBOR = do
+    language <- fromCBOR @Language
+    costModel <- decodeCostModel language
+    exUnits <- fromCBOR
+    sbs <- fromCBOR
+    pData <- fromCBOR
+    protVer <- fromCBOR
+    pure $ mkPlutusDebug isLanguage costModel exUnits sbs pData protVer
+
+data PlutusDebugWrapper where
+  PlutusDebugV1 :: PlutusDebug 'PlutusV1 -> PlutusDebugWrapper
+  PlutusDebugV2 :: PlutusDebug 'PlutusV2 -> PlutusDebugWrapper
+
+deriving instance Eq PlutusDebugWrapper
+
+deriving instance Show PlutusDebugWrapper
+
+deriving instance Generic PlutusDebugWrapper
+
+mkPlutusDebugWrapper ::
+  forall (l :: Language).
+  SLanguage l ->
+  PlutusDebug l ->
+  PlutusDebugWrapper
+mkPlutusDebugWrapper = \case
+  SPlutusV1 -> PlutusDebugV1
+  SPlutusV2 -> PlutusDebugV2
+
+instance ToCBOR PlutusDebugWrapper where
+  toCBOR = \case
+    PlutusDebugV1 pdbg -> toCBOR pdbg
+    PlutusDebugV2 pdbg -> toCBOR pdbg
+
+data PlutusError
+  = PlutusErrorV1 PV1.EvaluationError
+  | PlutusErrorV2 PV2.EvaluationError -- TODO: Should this also be made a GADT?
   deriving (Show)
 
-data PlutusDebugInfo
-  = DebugSuccess PV1.ExBudget
+data PlutusDebugInfo l
+  = DebugSuccess PV1.ExBudget -- NOTE: PV1.ExBudget == PV2.ExBudget, hence this works
```

Comment:
Same thing with this one. Maybe submitting a PR to plutus repo that exports this type from `PlutusLedgerApi.Common`? We can ask @michaelpj if there are expectations for `Data` and `ExBudget` to change with future plutus versions.

Reply:
Yes, that seems sensible, @bezirg could you look into that?

Reply:
Yes, I will export both Data and ExBudget from Common and will reply back here

Reply:
And `CostModelInterface (CostModelApplyWarn)` and `PlutusPrelude.enumerate` too :smile: . I have a branch open while I'm currently awaiting `nix develop` to finish, so I can check the build and push.